# 多线程下的异常处理

异步流程里最难处理也最容易忘记处理的是一些**异常情况**，比如超时。所以当我们在使用诸如 CompletableFuture 和 Listenablefuture 等带有异步属性和功能的API的时候，要注意对于超时的处理。
以 CompletableFuture 为例，如何保证方法在使用该类的异步API的基础上（不包括使用同步get等方法），对于超时能够做出有效的响应和处理。



## 多线程环境下异常被吞掉的情况

### Case1：模拟异常被分支线程吞掉的情况

```java
	/**
     * 模拟异常被分支线程吞掉的情况
     */
    @Test
    public void testSingleThread() {

        SmallTool.printTimeAndThread("主线任务开始执行");
        SmallTool.printTimeAndThread("创建一个分支线程");

        // 开启支线任务
        new Thread(() -> {
            SmallTool.printTimeAndThread("支线任务开始执行");
            // 任务执行时间
            SmallTool.sleepMillis(1000);
            // 模拟异常
            int i = 1 / 0;
            SmallTool.printTimeAndThread("支线任务执行成功");
        }).start();
        // 同步执行主线任务
        SmallTool.sleepMillis(10);
        SmallTool.printTimeAndThread("主线任务结束");

        /* 执行以上代码，输出结果为：
        --------------------------------------
        1658289136963	|	1	|	main	|	主线任务开始执行
        1658289136963	|	1	|	main	|	创建一个分支线程
        1658289137003	|	12	|	Thread-0	|	支线任务开始执行
        1658289137115	|	1	|	main	|	主线任务结束
        -----------------------------------
        分析：
        */
    }
```

我们可以看到，虽然上面这段代码出现了异常，由于异常发生在支线任务中，并不会影响主线任务，所以主线正常结束。

但对于支线任务来说，虽然启动了任务，但是由于异常导致任务未能成功。正常情况下，异常应该抛出错误信息让程序员进行排查，出现这种情况就是我们所说的『异常被吞掉了』



### Case2：测试『线程池+Runable』吞掉异常

```java
	@Test
    public void testExecute() {
        SmallTool.printTimeAndThread("主线任务开始执行");
        SmallTool.printTimeAndThread("创建一个分支线程");
        // 开启支线任务
        threadPool.execute(() -> {
            SmallTool.printTimeAndThread("支线任务开始执行");
            int i = 1 / 0;
            SmallTool.printTimeAndThread("支线任务执行成功");
        });
        // 同步执行主线任务
        SmallTool.sleepMillis(10);
        SmallTool.printTimeAndThread("主线任务结束");
        /* 执行以上代码，输出结果为：
        --------------------------------------
        1658289746344	|	1	|	main	|	主线任务开始执行
        1658289746344	|	1	|	main	|	创建一个分支线程
        1658289746384	|	12	|	pool-1-thread-1	|	支线任务开始执行
        Exception in thread "pool-1-thread-1" java.lang.ArithmeticException: / by zero
            at future.exceptionHandler.OldCase.lambda$test$1(OldCase.java:67)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
            at java.lang.Thread.run(Thread.java:748)
        1658289746409	|	1	|	main	|	主线任务结束
        -----------------------------------
        */
    }
```

同样，测试用例可以正常执行成功。虽然这次由于线程池的默认配置异常处理策略会将子线程的异常打印在控制台上，但是并未做保存、输出等操作。



### Case3：测试『线程池+Callable』吞掉异常

```java
    @Test
    public void testSubmit() {
        SmallTool.printTimeAndThread("主线任务开始执行");
        SmallTool.printTimeAndThread("创建一个分支线程");
        // 开启支线任务
        threadPool.submit(() -> {
            SmallTool.printTimeAndThread("支线任务开始执行");
            int i = 1 / 0;
            SmallTool.printTimeAndThread("支线任务执行成功");
            return i;
        });
        // 同步执行主线任务
        SmallTool.sleepMillis(10);
        SmallTool.printTimeAndThread("主线任务结束");
        /* 执行以上代码，输出结果为：
        --------------------------------------
        1658289946838	|	1	|	main	|	主线任务开始执行
        1658289946838	|	1	|	main	|	创建一个分支线程
        1658289946875	|	12	|	pool-1-thread-1	|	支线任务开始执行
        1658289946887	|	1	|	main	|	主线任务结束
        -----------------------------------
        分析：同样，测试用例执行成功，并且控制台不打印任何异常信息。异常被吞得干干净净。
        */
    }
```



----

## 如何处理被分支线程吞掉的异常

在生产环境中，异常日志是非常重要的，如果不能将系统中出现的异常保存到日志中，那么在定位、排查问题时将无从下手，因此，多线程环境下对分支线程异常的处理尤为重要。



### PlanA：覆写Thread对象的异常处理器UncaughtExceptionHandler

JDK为每个 Thread 对象提供了一个异常处理器 UncaughtExceptionHandler，如果你没有声明自定义的 UncaughtExceptionHandler 的话，就会使用 JDK 默认的。

```java
    @FunctionalInterface
    public interface UncaughtExceptionHandler {
        /**
         * 当给定线程由于未捕的获异常而终止时调用的方法。
         * Method invoked when the given thread terminates due to the
         * given uncaught exception.
         * <p>Any exception thrown by this method will be ignored by the
         * Java Virtual Machine.
         * @param t the thread
         * @param e the exception
         */
        void uncaughtException(Thread t, Throwable e);
    }
```

如果你在新创建的分支线程中覆写义这个方法，将自己的异常处理逻辑声明在其中，那么在发生异常时，就可以在这个方法中将日志打印输出到控制台上或者将日志写出到文件中。

具体代码如下：

```java
    /**
     * 通过复写Thread对象的异常处理器UncaughtExceptionHandler处理分支线程的异常
     */
    @Test
    public void test() {
        SmallTool.printTimeAndThread("主线任务开始执行");
        SmallTool.printTimeAndThread("创建一个分支线程");

        // 开启支线任务
        Thread t = new Thread(() -> {
            SmallTool.printTimeAndThread("支线任务开始执行");
            // 模拟异常
            int i = 1 / 0;
            SmallTool.printTimeAndThread("支线任务执行成功");
        });

        // 自定义异常处理器
        t.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
            @Override
            public void uncaughtException(Thread t, Throwable e) {
                SmallTool.printTimeAndThread(String.format("输出错误日志：线程【%s】出现了【%s】错误", t.getId(), e.getMessage()));
            }
        });

        t.start();
        /* 执行以上代码，输出结果为：
        --------------------------------------
        1658308391426	|	1	|	main	|	主线任务开始执行
        1658308391426	|	1	|	main	|	创建一个分支线程
        1658308391462	|	12	|	Thread-0	|	支线任务开始执行
        1658308391462	|	12	|	Thread-0	|	输出错误日志：线程【12】出现了【/ by zero】错误
        -----------------------------------
        分析：
        */
    }
```

当然，我们也可以稍微优化一些这段代码，将配置异常处理器UncaughtExceptionHandler的进行统一管理。

1. 创建一个自定义的 Thread 对象，继承 Thread 类，并配置自定义的 UncaughtExceptionHandler。

   ```java
   class MyThread extends Thread {
       public MyThread(Runnable target) {
           super(target);
           this.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
               @Override
               public void uncaughtException(Thread t, Throwable e) {
                   System.out.println(String.format("输出错误日志：线程【%s】出现了【%s】错误", t.getId(), e.getMessage()));
               }
           });
       }
   }
   ```

   

2. 在项目代码中使用自定义的 Thread 对象开启线程。

   ```java
       @Test
       public void test2() {
           SmallTool.printTimeAndThread("主线任务开始执行");
           SmallTool.printTimeAndThread("创建一个分支线程");
   
           // 开启支线任务
           new MyThread(() -> {
               SmallTool.printTimeAndThread("支线任务开始执行");
               // 模拟异常
               int i = 1 / 0;
               SmallTool.printTimeAndThread("支线任务执行成功");
           }).start();
   
           // 同步执行主线任务
           SmallTool.sleepMillis(10);
           SmallTool.printTimeAndThread("主线任务结束");
   
           /* 执行以上代码，输出结果为：
           --------------------------------------
           1658308689156	|	1	|	main	|	主线任务开始执行
           1658308689156	|	1	|	main	|	创建一个分支线程
           1658308689197	|	12	|	Thread-0	|	支线任务开始执行
           1658308689197	|	12	|	Thread-0	|	输出错误日志：线程【12】出现了【/ by zero】错误
           1658308689216	|	1	|	main	|	主线任务结束
           -----------------------------------
           分析：
           */
       }
   ```

   

-------

### PlanB：使用 ListenableFuture 提供的异常处理API

ListenableFuture 是 Guava 提供的用于处理多线程异步任务的 API，里面提供了多种方法用于异常处理。

使用 ListenableFuture 处理分支线程的异常代码如下：

```java
package future.listenableFuture;

import com.google.common.util.concurrent.*;
import utils.SmallTool;

import java.util.concurrent.Executors;

/**
 * @Author lnd
 * @Description
 * @Date 2022/7/20 10:30
 */
public class FutureHandleExceptionTest {
    /**
     * 线程池中线程的个数
     */
    private static final int POOL_SIZE = 50;

    /**
     * 带有回调机制的线程池
     * （装饰器模式）
     */
    private static ListeningExecutorService listeningExecutorService = MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(POOL_SIZE));

    public static void main(String[] args) {
        SmallTool.printTimeAndThread("主线任务开始执行");

        /**
         * 将实现了 Callable 的任务放到线程池中，得到一个带有回调机制的 ListenableFuture 实例（获取线程执行结果的凭据）。
         * 通过 Futures.addCallback 方法对得到的 ListenableFuture 实例进行监听，一旦得到结果（正常执行结果/异常执行结果）.
         * 正常结果进入 onSuccess 方法中，异常结果进入 onFailure 方法处理。
         */
        ListenableFuture<String> listenableFuture = listeningExecutorService
                // 通过（带监听器的）线程池执行任务
                .submit(() -> {
                    SmallTool.printTimeAndThread("支线任务开始执行");
                    // 执行任务耗时
                    SmallTool.sleepMillis(3000);
                    // 模拟异常情况
                    int i = 1 / 0;
                    // 执行任务成功
                    SmallTool.printTimeAndThread("支线任务执行成功");
                    return "OK";
                });

        SmallTool.printTimeAndThread("主线任务继续运行");

        // 添加监听器对 listenableFuture 实例进行监听（运行监听器需要一个额外的线程，由线程池进行分配）
        listenableFuture.addListener(() -> SmallTool.printTimeAndThread("监听器正在监听任务的执行情况"), listeningExecutorService);

        // 通过 listenableFuture 凭据获取执行结果
        Futures.addCallback(listenableFuture, new FutureCallback<String>() {
            /**
             * 处理正常结果
             * @param result
             */
            public void onSuccess(String result) {
                SmallTool.printTimeAndThread("程序执行完成，执行结果为：" + result);
            }

            /**
             * 处理异常结果
             * @param t
             */
            public void onFailure(Throwable t) {
                SmallTool.printTimeAndThread("程序出现异常，异常信息为：" + t.getMessage());
                //throw new RuntimeException(t);

            }
        });

        // 同步执行主线任务
        SmallTool.sleepMillis(100);
        SmallTool.printTimeAndThread("主线任务结束");

        /* 执行以上代码，输出结果为：
        --------------------------------------
            1658309255518	|	1	|	main	|	主线任务开始执行
            1658309255577	|	1	|	main	|	主线任务继续运行
            1658309255578	|	13	|	pool-1-thread-1	|	支线任务开始执行
            1658309258594	|	15	|	pool-1-thread-2	|	监听器正在监听任务的执行情况
            1658309258594	|	13	|	pool-1-thread-1	|	程序出现异常，异常信息为：/ by zero
        -----------------------------------
        分析：
        */
    }
}
```



### PlanC：使用 CompletableFuture 提供的异常处理 API

CompletableFuture 是 JDK8 新增的用于处理异步线程的 API，其提供了一系列异常处理 API 也可以用于处理分支线程中的异常。

场景如下：

```markdown
# 题目：
异步流程里最难处理也最容易忘记处理的是一些**异常情况**，比如超时。
所以当我们在使用诸如CompletableFuture和Listenablefuture等带有异步属性和功能的API的时候，要注意对于超时的处理。
以CompletableFuture为例，如何保证方法在使用该类的异步API的基础上（不包括使用同步get等方法），对于超时能够做出有效的响应和处理。

# 分析
对于程序中某个线程运行超时，主要存在两种情况：
- Case1：线程正常运行，只是确实这个线程正在执行的任务耗时非常长；
- Case2：线程遇到异常情况，已经挂了。

# 解决方案：
## Case1
1. 我们可以额外添加一个“守护线程”（假设为线程B），这个守护线程执行的时间是我们可以容忍正常线程（假设为线程A）运行的最大时间。 
2. 同时开启这两个异步线程： 
   1. 正常情况下，线程A的运行时间肯定要短于线程B，所以它的执行结果一定先于线程返回；
   2. 在特殊情况下，比如线程由于网络、或出现异常等原因超时了，那么线程B的执行结果就会先于线程A返回。
3. 如果我们从代码中得到了线程B的返回结果，那我们就知道线程A出现问题了。

### 优/缺点：
- 优点：这个方案的优点是：可能有些时候，程序确实是正常运行的，只是处理得比较慢，这个时候程序是不会抛出异常的，只能通过这种方式来进行处理。
- 缺点：但是这个方案的缺点就是：我们只能知道线程A出现问题了，却并不能知道线程A出现了什么问题。


## Case2
对于真正出现异常的情况，可以利用 CompletableFuture 提供的异常处理机制。使用 exceptionally、handle 和 whenComplete 三种方法之中的任一即可。
```



具体代码如下：

```java
package future.completableFuture.handleException;

import utils.SmallTool;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

/**
 * @Author lnd
 * @Description
 * @Date 2022/7/19 18:58
 */
public class TimeoutHandler {
    /**
     * 程序超时时间
     */
    private Long timeout;

    /**
     * 程序执行时间
     */
    private Long executeTime;

    public TimeoutHandler(Long timeout, Long executeTime) {
        this.timeout = timeout;
        this.executeTime = executeTime;
    }

    /**
     * 线程池
     */
    public ThreadPoolExecutor threadPool = new ThreadPoolExecutor(
            3,
            Integer.MAX_VALUE,
            0,
            TimeUnit.SECONDS,
            new SynchronousQueue<>());

    public void start() {
        CompletableFuture<Boolean> cf1 = CompletableFuture
                // 开启异步线程执行任务
                .supplyAsync(() -> {
                    boolean flag = false;
                    SmallTool.printTimeAndThread("数据传输中，预估耗时" + executeTime + "毫秒");
                    // 任务执行时间
                    SmallTool.sleepMillis(executeTime);

                    // 模拟异常
                    int i = 1 / 0;

                    // 任务执行成功
                    flag = true;
                    return flag;
                })
                // 同时开启一个“守护线程”
                .applyToEither(
                        CompletableFuture.supplyAsync(() -> {
                            // 守护线程什么也不干，只设置允许等待的最大时间
                            SmallTool.printTimeAndThread("我是守护线程，我最多只等" + timeout + "毫秒");
                            SmallTool.sleepMillis(timeout);
                            return false;
                        }), value -> value
                )
                .exceptionally(ex -> {
                    // 处理异常
                    SmallTool.printTimeAndThread("Exception" + ex.getMessage());
                    return null;
                });

        if (cf1.join()){
            SmallTool.printTimeAndThread("程序执行成功");
        }else {
            SmallTool.printTimeAndThread("程序执行超时");
        }
    }


    public static void main(String[] args) {
        TimeoutHandler timeoutHandler = new TimeoutHandler(3000L, 2000L);
        //TimeoutHandler timeoutHandler = new TimeoutHandler(2000L, 3000L);
        timeoutHandler.start();

        /* 执行超时的情况：
        --------------------------------------
        1658232274797	|	12	|	ForkJoinPool.commonPool-worker-1	|	数据传输中，预估耗时2000毫秒
        1658232274797	|	13	|	ForkJoinPool.commonPool-worker-2	|	我是守护线程，我最多只等1000毫秒
        1658232275810	|	1	|	main	|	程序执行超时
        ----------------------------------- */

        /* 执行成功的情况：
        --------------------------------------
        1658232292422	|	13	|	ForkJoinPool.commonPool-worker-2	|	我是守护线程，我最多只等5000毫秒
        1658232292422	|	12	|	ForkJoinPool.commonPool-worker-1	|	数据传输中，预估耗时2000毫秒
        1658232294427	|	1	|	main	|	程序执行成功
        ----------------------------------- */
        
        /* 出现异常的情况：
        --------------------------------------
        1658309932444	|	13	|	ForkJoinPool.commonPool-worker-2	|	我是守护线程，我最多只等3000毫秒
        1658309932444	|	12	|	ForkJoinPool.commonPool-worker-1	|	数据传输中，预估耗时2000毫秒
        1658309934460	|	12	|	ForkJoinPool.commonPool-worker-1	|	Exceptionjava.lang.ArithmeticException: / by zero
        ----------------------------------- 
        */
    }
}
```





